utils/wasm-instrument/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+            .internal()
+            .global(gas_index)
+            .build(),
+    );
+
+    mbuilder.push_global(
+        builder::global()
+            .value_type()
+            .i64()
+            .init_expr(Instruction::I64Const(0))
+            .mutable()
+            .build(),
+    );
+
+    mbuilder.push_export(
+        builder::export()
+            .field(GLOBAL_NAME_ALLOWANCE)
+            .internal()
+            .global(allowance_index)
+            .build(),
+    );
+
+    let mut elements = vec![
+        // check if there is enough gas
+        Instruction::GetGlobal(gas_index),
+        // calculate gas_to_charge + cost_for_func
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // if gas < (gas_to_charge + cost_for_func)
+        Instruction::I64LtU,
+        Instruction::If(elements::BlockType::NoResult),
+        Instruction::Call(out_of_gas_index),
+        Instruction::Unreachable,
+        Instruction::End,
+        // update gas
+        Instruction::GetGlobal(gas_index),
+        // calculate gas_to_charge + cost_for_func
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // gas -= (gas_to_charge + cost_for_func)
+        // {
+        Instruction::I64Sub,
+        Instruction::SetGlobal(gas_index),
+        // }
+        // check if there is enough gas allowance
+        Instruction::GetGlobal(allowance_index),
+        // calculate gas_to_charge + cost_for_func
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // if allowance < (gas_to_charge + cost_for_func)
+        Instruction::I64LtU,
+        Instruction::If(elements::BlockType::NoResult),
+        Instruction::Call(out_of_allowance_index),
+        Instruction::Unreachable,
+        Instruction::End,
+        // update gas allowance
+        Instruction::GetGlobal(allowance_index),
+        // calculate gas_to_charge + cost_for_func
+        // {
+        Instruction::GetLocal(0),
+        Instruction::I64ExtendUI32,
+        Instruction::I64Const(i64::MAX),
+        Instruction::I64Add,
+        // }
+        // allowance -= (gas_to_charge + cost_for_func)
+        // {
+        Instruction::I64Sub,
+        Instruction::SetGlobal(allowance_index),
+        // }
+        Instruction::End,
+    ];
+
+    // determine cost for successful execution
+    let cost_check_blocks = match elements.iter().take(7).try_fold(0u64, |cost, instruction| {
+        rules.instruction_cost(instruction).map(|c| cost + c as u64)
+    }) {
+        Some(c) => 2 * c,
+        None => return Err(mbuilder.build()),
+    };
+
+    let cost_set_blocks = match elements
```

Comment:
Why do we need to calculate `cost_check_blocks` and `cost_set_blocks` separately if we use them like this `cost_check_blocks + cost_set_blocks`

Reply:
Since instructions inside "if" block are skipped: they executed only on out of gas/allowance

Reply:
What about this?
```rust
    let cost = elements
        .iter()
        .take(7)
        .chain(elements.iter().skip(10).take(7))
        .try_fold(0u64, |cost, instruction| {
            rules.instruction_cost(instruction).map(|c| cost + c as u64)
        });
```

Reply:
Also can you make checked add: `|c| cost + c as u64`. And make check after that the result `cost` for gas_charge function is less then u32::MAX

Reply:
good remark! I didn't so 'cause my eyes are full of soap =)